src/main/scala/play/api/db/slick/internal/DBApiAdapter.scala

```diff
@@ -0,0 +1,57 @@
+package play.api.db.slick.internal
+
+import java.sql.Connection
+
+import javax.inject.Inject
+import javax.sql.DataSource
+
+import play.api.db.DBApi
+import play.api.db.{Database => PlayDatabase}
+import play.api.db.slick.DbName
+import play.api.db.slick.SlickApi
+
+import slick.backend.DatabaseConfig
+import slick.driver.JdbcProfile
+
+private[slick] class DBApiAdapter @Inject() (slickApi: SlickApi) extends DBApi {
+  private lazy val databasesByName: Map[DbName, PlayDatabase] = slickApi.dbConfigs[JdbcProfile]().map {
+    case (name, dbConfig) => (name, new DBApiAdapter.DatabaseAdapter(name, dbConfig))
+  }(collection.breakOut)
+
+  override def databases: Seq[PlayDatabase] = databasesByName.values.toSeq
+
+  def database(name: String): PlayDatabase =
+    databasesByName.getOrElse(DbName(name), throw new IllegalArgumentException(s"Could not find database for $name"))
+
+  def shutdown(): Unit = {
+    // no-op: shutting down dbs is automatically managed by `slickApi`
+    ()
+  }
+}
+
+private[slick] object DBApiAdapter {
+  private class DatabaseAdapter(_name: DbName, dbConfig: DatabaseConfig[JdbcProfile]) extends PlayDatabase {
+    def name: String = _name.value
+    def dataSource: DataSource = throw new UnsupportedOperationException
```

Comment:
You can look at `.db.source` but there's no common abstraction, so you'll have to handle `DataSourceJdbcDataSource` and `HikariCPJdbcDataSource` separately (both have a `ds` field). We could make the latter a subclass of the former in 3.1.